### PR TITLE
Parse byte array of otherName SANs to string of ASN1 object and OID

### PR DESCRIPTION
**Pull request by jksmth**
Signed-off-by: Jake Smith <jakemgsmith@gmail.com>

##  opendistro-for-elasticsearch/security pull request intake form
 
 1. __Category:__ Bug fix
 
 
 
 2. __Github Issue__ #843

 3. __Description of changes:__
Checks for type 0 (otherName) Subject Alternate Names (SANs) when validating certificate DNs, if found the byte array is read and if is a valid ASN1 object the OID and value is decoded to provide a string representation of the otherName. This is achieved with the use of bouncycastle lib.

 4. __Why these changes are required?__ 
 Per #843 `cert.getSubjectAlternativeNames().toString()` does not lookup type 0 (otherName) SANs correctly and returns a string representing the byte array which changes on each iteration. This causes hot reload to fail if the certificate includes an otherName in the SAN as the string of the SANs returned by `cert.getSubjectAlternativeNames().toString()` will never match the current and new certificate in the `hasValidDNs()` method.

 5. __What is the old behavior before changes and new behavior after changes?__

__Old behaviour__
Getting currently loaded certificates with `GET /_opendistro/_security/api/ssl/certs` would return:
```
{
  "http_certificates_list" : [
    {
      "issuer_dn" : "CN=elastic-test-http,OU=elastic-test",
      "subject_dn" : "CN=elastic-test-es-http.local,OU=elastic-test",
      "san" : "[[2, elastic-test-es-http]]",
      "not_before" : "2020-11-19T12:42:57Z",
      "not_after" : "2020-11-19T12:57:57Z"
    }
  ],
  "transport_certificates_list" : [
    {
      "issuer_dn" : "CN=elastic-test-transport,OU=elastic-test",
      "subject_dn" : "CN=elastic-test-es-node-0.local,OU=elastic-test",
      "san" : "[[0, [B@6ac191ff], [2, elastic-test-es-node-0], [7, 127.0.0.1]]",
      "not_before" : "2020-11-19T11:48:35Z",
      "not_after" : "2020-11-19T12:03:35Z"
    }
  ]
}
```

Note `[0, [B@6ac191ff]` in transport certificate san. Repeated calls to GET /_opendistro/_security/api/ssl/certs shows that the string representation of the OtherName byte array (B@xxxxxxxx) changes each time.

__New behaviour__
Getting currently loaded certificates with `GET /_opendistro/_security/api/ssl/certs` will now return:
```
{
  "http_certificates_list" : [
    {
      "issuer_dn" : "CN=elastic-test-http,OU=elastic-test",
      "subject_dn" : "CN=elastic-test-es-http.local,OU=elastic-test",
      "san" : "[[2, elastic-test-es-http]]",
      "not_before" : "2020-11-19T12:42:57Z",
      "not_after" : "2020-11-19T12:57:57Z"
    }
  ],
  "transport_certificates_list" : [
    {
      "issuer_dn" : "CN=elastic-test-transport,OU=elastic-test",
      "subject_dn" : "CN=elastic-test-es-node-0.local,OU=elastic-test",
      "san" : "[[0, [2.5.4.3, elastic-test-es-node-0]], [2, elastic-test-es-node-0], [7, 127.0.0.1]]",
      "not_before" : "2020-11-19T11:48:35Z",
      "not_after" : "2020-11-19T12:03:35Z"
    }
  ]
}
```
Note `[0, [2.5.4.3, elastic-test-es-node-0]]` in transport certificate san is now decoded to a list toString() of [oid, value]

 6. __Testing done:__
 Verified by manually testing and unit tests. New test certificates need creating to include an otherName in the SAN.  
 
 
 7. __TO-DOs, if any:__
 

 8. __Is it backport from master branch?__
 No

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by jksmth** [comment]
To update the unit tests, certificates and keystrores need updating in `src/test/resources/ssl/reload` to include otherName in the SAN, is there a script or openssl config file that is used for generating these?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/DefaultOpenDistroSecurityKeyStore.java [hunk 1]
```diff
@@ -950,4 +976,36 @@ private static void checkPath(String keystoreFilePath, String fileNameLogOnly) {
...
+        ASN1Sequence seq = ASN1Sequence.getInstance(oct);
+
+        // Get object identifier from first in sequence
+        ASN1ObjectIdentifier asnOID = (ASN1ObjectIdentifier) seq.getObjectAt(0);
+        String oid = asnOID.getId();
+
+        // Get value of object from second element
+        final ASN1TaggedObject obj = (ASN1TaggedObject) seq.getObjectAt(1);
+        // Could be tagged twice due to bug in java cert.getSubjectAltName
+        ASN1Primitive prim = obj.getObject();
```
Comment (debjanibnrj) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/DefaultOpenDistroSecurityKeyStore.java:1000]: slightly confused by what this code section does, can you mention the bug you're referencing.

**Review by debjanibnrj — COMMENTED** [review]
Thanks for your contribution @jksmth! 

You're solution looks good, could you add unit tests to this as well and submit as a PR.

**Comment by debjanibnrj** [comment]
Hey @jksmth, Friendly ping to see if you need any help with this. Would like to work with you to get this in.

**Comment by jksmth** [comment]
Hello, yes happy to get this finished off. There are quite a few test certs that need generating for the unit tests. I'll try and make a start on it in the next few days. Any helper scripts around to help generate them or is it a manual task?

**Comment by debjanibnrj** [comment]
The cert generation tool mentioned here - https://docs.search-guard.com/latest/offline-tls-tool#tls-tool should help.

**Comment by debjanibnrj** [comment]
Hey @jksmth, Looks like this branch needs to be rebased with main

**Comment by jksmth** [comment]
Hello @debjanibnrj I have rebased and had to force-push and close/reopen this PR to get it to pickup the changes following the source repo rename to OpenSearch. Please approve the workflow if you are happy and let me know any comments, functionality and tests are now complete.

The certs were generated with the offline-tls-tool as you suggested - FYI I also have a PR into that project to add otherName support:  https://github.com/floragunncom/search-guard-tlstool/pull/18

**Review by debjanibnrj — APPROVED** [review]
Thanks for your contribution! 

@hardik-k-shah @vrozov Let me know if you can have a look

**Comment by jksmth** [comment]
Hello

Are you still interested in this PR? I notice my branch is behaving a bit strangely in GitHub as it was forked before the project rename.

Would you like me to refork, rebase and submit a new PR?

**Comment by hardik-k-shah** [comment]
> Hello
> 
> Are you still interested in this PR? I notice my branch is behaving a bit strangely in GitHub as it was forked before the project rename.
> 
> Would you like me to refork, rebase and submit a new PR?

Yes, please resubmit it on OpenSearch repo. Thanks.

**Comment by jksmth** [comment]
Closing this PR - replaced by #1369

**Pull request closed without merging**